### ISSUE TITLE
twisted.python.modules does not account for memory/disk inconsistencies when scanning packages

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#2643]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)| @glyph reported|
|-|-|
|Trac ID|trac#2643|
|Type|enhancement|
|Created|2007-05-14 18:25:44Z|

This script:
```
#!py
from twisted.python.modules import walkModules
import mx.DateTime
print list(walkModules())
```
results in this error:
```
Traceback (most recent call last):
  File "fbi.py", line 6, in <module>
    print list(walkModules())
  File "/home/glyph/Projects/Twisted/trunk/twisted/python/modules.py", line 729, in walkModules
    for module in package.walkModules(importPackages=False):
  File "/home/glyph/Projects/Twisted/trunk/twisted/python/modules.py", line 174, in walkModules
    for module in package.walkModules(importPackages=importPackages):
  File "/home/glyph/Projects/Twisted/trunk/twisted/python/modules.py", line 174, in walkModules
    for module in package.walkModules(importPackages=importPackages):
  File "/home/glyph/Projects/Twisted/trunk/twisted/python/modules.py", line 173, in walkModules
    for package in self.iterModules():
  File "/home/glyph/Projects/Twisted/trunk/twisted/python/modules.py", line 125, in iterModules
    for placeToLook in self._packagePaths():
  File "/home/glyph/Projects/Twisted/trunk/twisted/python/modules.py", line 418, in _packagePaths
    for fn in self.load().__path__:
AttributeError: 'module' object has no attribute '__path__'
```

```_packagePaths``` wrongly assumes that any object which appears to be a package on disk will actually load a package object in memory, which is wrong for a few reasons:
  * multiple sys.path entries can confuse the module loader
  * a module can actually end up being anything it wants from ```namedAny```, so you can't depend on any particular fixed contract
  * even "real" packages might delete their ```__path__``` for some reason

In mxDateTime's case, the module is extremely poorly behaved (it's rife with ```import *``` statements, two of which are nested, resulting in a package clobbering its own name), but the general problem is still tpm's fault.

This manifests in a test failure with the above traceback if you have pypgsql and mxDateTime installed and you run the full suite.  It doesn't manifest otherwise because tpm never tries to load the mxDateTime packages otherwise.

<details><summary>Searchable metadata</summary>

```
trac-id__2643 2643
type__enhancement enhancement
reporter__glyph glyph
priority__highest highest
milestone__ 
branch__ 
branch_author__ 
status__closed closed
resolution__fixed fixed
component__core core
keywords__ 
time__1179167144000000 1179167144000000
changetime__1193591079000000 1193591079000000
version__None None
owner__therve therve
cc__therve
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_2"></a>|@therve set owner to @radix|
|-|-|

This is maybe fixed in modules-path-2643. At least the reported problem doesn't happen anymore, but I'm not sure I understand this enough.

radix, I'm sure you'd love to review this.

**Comment by twisted-trac** [comment]
|[<img alt="radix's avatar" src="https://avatars.githubusercontent.com/u/227068?s=50" width="50" height="50">](https://github.com/radix)<a name="note_3"></a>|@radix set owner to @therve|
|-|-|

I guess this is OK.

Really this whole situation is bad. Here's my explanation of the problem: since mx clobbers its own namespace, depending on how or when you import it you might get different objects. Here's how this is affecting t.p.m.

TRUNK:

```
>>> from twisted.python.modules import getModule
>>> print list(getModule('mx.DateTime.mxDateTime').walkModules())
[PythonModule<'mx.DateTime.mxDateTime'>, ... BIG LIST OF MODULES]
>>> import mx.DateTime.mxDateTime
>>> print list(getModule('mx.DateTime.mxDateTime').walkModules())
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "twisted/python/modules.py", line 173, in walkModules
    for package in self.iterModules():
  File "twisted/python/modules.py", line 125, in iterModules
    for placeToLook in self._packagePaths():
  File "twisted/python/modules.py", line 418, in _packagePaths
    for fn in self.load().__path__:
AttributeError: 'module' object has no attribute '__path__'
```

Your branch::

```
>>> from twisted.python.modules import getModule
>>> print list(getModule('mx.DateTime.mxDateTime').walkModules())
[PythonModule<'mx.DateTime.mxDateTime'>, ...SAME BIG LIST OF MODULES]
>>> import mx.DateTime.mxDateTime
>>> print list(getModule('mx.DateTime.mxDateTime').walkModules())
[PythonModule<'mx.DateTime.mxDateTime'>]
```

I think this is acceptable, since walkModules will reflect what is importable at the time that it's called.

+1

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_4"></a>|@therve set status to `closed`|
|-|-|

(In [21452]) Merge modules-path-2643

Author: therve
Reviewer: radix
Fixes [#2643](https://github.com/twisted/twisted/issues/2643)

Manage packages which delete their __path__ attribute in
twisted.python.modules.walkModules and twisted.python.modules.iterModules.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_5"></a>|@therve commented|
|-|-|

(In [21471]) Revert r21452: regression where tests are run in another order

Refs [#2643](https://github.com/twisted/twisted/issues/2643)

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_6"></a>|@therve set status to `reopened`|
|-|-|

This was a stupid here in test added, fixed in [21470].

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)|@therve set owner to @radix<br>@therve set status to `new`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="radix's avatar" src="https://avatars.githubusercontent.com/u/227068?s=50" width="50" height="50">](https://github.com/radix)<a name="note_8"></a>|@radix set owner to @therve|
|-|-|

Sorry for missing that the first time around. Yes, hooray global state. +1.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_10"></a>|@therve set status to `closed`|
|-|-|

(In [21472]) Merge modules-path-2643

Author: therve
Reviewer: radix
Fixes [#2643](https://github.com/twisted/twisted/issues/2643)

Manage packages which delete their __path__ attribute in
twisted.python.modules.walkModules and twisted.python.modules.iterModules.
